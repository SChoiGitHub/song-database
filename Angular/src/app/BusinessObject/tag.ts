export type RawTag = {
  tagId: number,
  name: string,
  description: string,
}

export class Tag {
  tagId: number;
  name: string;
  description: string;

  constructor(rawData: Partial<RawTag>) {
    const {
      tagId = 0,
      name = '',
      description = '',
    } = rawData;

    this.tagId = tagId;
    this.name = name;
    this.description = description;
  }
}
