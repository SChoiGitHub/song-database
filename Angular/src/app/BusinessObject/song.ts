import { RawTag, Tag } from "./tag";

export type RawSong = {
  songId: number,
  name: string,
  length: number,
  tags: RawTag[],
}

export class Song {
  songId: number;
  name: string;
  length: number;

  tags: Tag[];

  constructor(rawData: Partial<RawSong> = {}) {
    const {
      songId = -1,
      name = '',
      length = 0,
      tags = [{ name: '' }],
    } = rawData;

    this.songId = songId;
    this.name = name;
    this.length = length;
    this.tags = tags.map(t => new Tag(t));
  }

  isAdding() {
    return this.songId === -1;
  }

  isEditing() {
    return this.songId !== -1;
  }
}
