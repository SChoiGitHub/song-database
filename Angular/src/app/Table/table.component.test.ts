import { expect } from '@jest/globals';
import { TestBed } from '@angular/core/testing';
import { TableComponent } from './table.component';
import { MusicService } from '../Service/music.service';
import { MusicServiceDouble } from '../Service/__mocks__/music.service';
import { Song } from '../BusinessObject/song';
import { BrowserTestingModule } from '@angular/platform-browser/testing';

describe('TableComponent', () => {
  let mock: MusicServiceDouble;

  beforeEach(() => {
    mock = new MusicServiceDouble();

    const providers = [
      { provide: MusicService, useValue: mock }
    ];

    TestBed.configureTestingModule({
      imports: [BrowserTestingModule],
      providers,
      declarations: [TableComponent]
    });
  });

  it('should say no music added without songs', () => {
    //Act
    const fixture = TestBed.createComponent(TableComponent);
    const app = fixture.nativeElement;

    //Assert
    expect(app).toMatchSnapshot()
  });

  it('should list songs', () => {
    //Arrange
    const song = new Song({ name: "Reborn Again", length: 209 });
    const fixture = TestBed.createComponent(TableComponent);

    //Act
    mock.songs.next([song]);
    fixture.detectChanges();
    const app = fixture.nativeElement;

    //Assert
    expect(app).toMatchSnapshot()
  });
});
