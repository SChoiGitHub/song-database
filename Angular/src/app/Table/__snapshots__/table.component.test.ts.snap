// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableComponent should list songs 1`] = `
<div
  id="root1"
>
  <h5>
     Music List

  </h5>
  <div
    class="songTable"
    ngbaccordion=""
  >
    <div
      ngbaccordionitem=""
    >
      <h2
        ngbaccordionheader=""
      >
        <button
          ngbaccordionbutton=""
        >
          Reborn Again
        </button>
      </h2>
      <div
        ngbaccordioncollapse=""
      >
        <div
          ngbaccordionbody=""
        >
          <ul>
            <li>
              <a>
                Edit
              </a>
            </li>
            <li>
              <a>
                Delete
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    
    
  </div>
  
  
  
</div>
`;

exports[`TableComponent should say no music added without songs 1`] = `
<div
  id="root0"
>
  <h5>
     Music List

  </h5>
  
  
</div>
`;
